### ISSUE TITLE
Deduced actual type of variable not used in lambda context

### ISSUE THREAD

**Issue by oldium** [microsoft/pylance-release#1016]
## Environment data

-   Language Server version: 2021.2.4 (pyright a18e6fb4)
-   OS and version: Windows 10
-   Python version (& distribution if applicable, e.g. Anaconda): Python 3.4.1, typing 3.6.2 (from PyPi)

## Expected behaviour

Deduced actual type (reduced by isinstance's) is recognised in lambda context.

## Actual behaviour

When typed argument is used inside lambda, the original argument type is used instead of the one reduced by isinstance's.

Warning:

```
(parameter) value: int | (*args: Unknown, **kwargs: Unknown) -> Any
Object of type "int" is not callablePylancereportGeneralTypeIssues)
```

## Logs

This is reported from a company notebook, so I cannot provide logs (data collection restrictions).

## Code Snippet / Additional information

This fails, see the last line and a `value()` call.:

```python
from typing import Callable
def int_to_value(value: 'int | Callable') -> 'dict':
  if isinstance(value, int):
    return dict(hex="0x{:X}".format(value))
  else:
    return dict(hex=lambda: "0x{:X}".format(value()))
```

Please note that the following works correctly:

```python
from typing import Callable
def int_to_value(value: 'int | Callable') -> 'dict':
  if isinstance(value, int):
    return dict(hex="0x{:X}".format(value))
  else:
    return dict(hex="0x{:X}".format(value()))
```

**Comment by erictraut** [comment]
Type narrowing is not applied to variables captured within a lambda context because captured symbols can be reassigned a new value (with a different type) by the time the lambda is executed. In this particular case, it is safe, but in general a type checker cannot make this assumption.

The workaround is to assign the narrowed value to a local variable that has a non-union callable type.

```python
from typing import Callable
def int_to_value(value: 'int | Callable') -> 'dict':
  if isinstance(value, int):
    return dict(hex="0x{:X}".format(value))
  else:
    callable_value = value
    return dict(hex=lambda: "0x{:X}".format(callable_value()))
```

**Comment by oldium** [comment]
Perfect, thanks for really fast response, it works fine with this workround.

**Comment by erictraut** [comment]
I found a way to handle type narrowing of local variables and parameters captured by inner-scoped functions and lambdas, so the above workaround will no longer be necessary in this case.

**Comment by jakebailey** [comment]
This issue has been fixed in version 2021.9.4, which we've just released. You can find the changelog here: https://github.com/microsoft/pylance-release/blob/main/CHANGELOG.md#202194-29-september-2021

**Comment by oldium** [comment]
Works great. Thanks!
